jetty-server/src/main/java/org/eclipse/jetty/server/AcceptRateLimit.java

```diff
@@ -0,0 +1,217 @@
+ *   ...
+ *   server.start();
+ * </pre>
+ * @see SelectorManager.AcceptListener
+ */
+@ManagedObject
+public class AcceptRateLimit extends AbstractLifeCycle implements SelectorManager.AcceptListener, Runnable
+{
+    private static final Logger LOG = Log.getLogger(AcceptRateLimit.class);
+    
+    private final Server _server;
+    private final List<AbstractConnector> _connectors = new ArrayList<>();
+    private final Rate _rate;
+    private final int _maxRate;
+    private boolean _limiting = false;
+
+    public AcceptRateLimit(@Name("maxRate") int maxRate, @Name("period") long period, @Name("units") TimeUnit units, @Name("server") Server server)
+    {
+        _server = server;
+        _maxRate = maxRate;
+        _rate = new Rate(period,units);
+    }
+    
+    public AcceptRateLimit(@Name("maxRate") int maxRate, @Name("period") long period, @Name("units") TimeUnit units, @Name("connectors") Connector...connectors)
+    {
+        _server = null;
+        _maxRate = maxRate;
+        _rate = new Rate(period,units); 
+        for (Connector c: connectors)
+        {
+            if (c instanceof AbstractConnector)
+                _connectors.add((AbstractConnector)c);
+            else
+                LOG.warn("Connector {} is not an AbstractConnection. Connections not limited",c);
+        }
+    }
+    
+    protected void age(long period, TimeUnit units)
+    {
+        _rate.age(period,units);
+    }
+    
+    @Override
+    protected synchronized void doStart() throws Exception
+    {
+        if (_server!=null)
+        {
+            for (Connector c: _server.getConnectors())
+            {
+                if (c instanceof AbstractConnector)
+                    _connectors.add((AbstractConnector)c);
+                else
+                    LOG.warn("Connector {} is not an AbstractConnection. Connections not limited",c);
+            }
+        }
+
+        if (LOG.isDebugEnabled())
+            LOG.debug("AcceptLimit accept<{} rate<{} in {} for {}", _maxRate, _rate,_connectors);
+        
+        _limiting = false;
+        
+        for (AbstractConnector c : _connectors)
+            c.addBean(this);
+    }
+
+    @Override
+    protected synchronized void doStop() throws Exception
+    {
+        for (AbstractConnector c : _connectors)
+            c.removeBean(this);
+        if (_server!=null)
+            _connectors.clear();   
+    }
+
+    protected void limit()
+    {
+        for (AbstractConnector c : _connectors)
+        {
+            c.setAccepting(false);
+        }
+        
+        Scheduler scheduler = _connectors.get(0).getScheduler();
+        scheduler.schedule(this,_rate.getPeriod(),_rate.getUnits());
+    }
+    
+    protected void unlimit()
+    {
+        for (AbstractConnector c : _connectors)
+        {
+            c.setAccepting(true);
+        }
+    }    
+
+    @Override
+    public synchronized void onAccepting(SelectableChannel channel)
```

Comment:
-1 on using `synchronized` as modifier. Use `synchronized` blocks as small as possible - seems to me here that the only field that needs guard is `_limiting`. Do you need `limit()` to be called while holding the lock ?

Reply:
I agree.... just that big recent client doesn't and I've been trained to put them on the methods by their code reviews :)  
Also, I should synchronize on the `_rate` instance so there is only a single lock.
I think `limit()` needs to be protected else a call to `unlimit()` may overtake a call to `limit()`